File: packages/platos-client-py/platos_client/apis/threads.py
"""Threads API + realtime streaming for platos-client.

Theme I.7 — `send(...)` is an async-generator over agent stream events.
Connection management mirrors the TS SDK:

* Manual reconnection with exponential backoff.
* Events during a disconnect gap are buffered by the server's per-thread
  room; the client re-joins and resumes the stream.
* `asyncio.CancelledError` tears down the socket and yields a terminal
  `done` event with `stopped=True`.
"""

from __future__ import annotations

import asyncio
import json
from typing import TYPE_CHECKING, Any, AsyncGenerator

import websockets
from websockets.asyncio.client import ClientConnection

from platos_client.errors import PlatosError, PlatosNetworkError, PlatosNotFoundError

if TYPE_CHECKING:
    from platos_client.client import PlatosClient, PlatosScope


_DEFAULT_MAX_RECONNECT = 5
_RECONNECT_BASE_S = 0.5
_RECONNECT_MAX_S = 10.0


class ThreadsApi:
    def __init__(self, client: "PlatosClient") -> None:
        self._client = client

    # ---- CRUD ----

    async def create(
        self,
        *,
        scope: "PlatosScope | None" = None,
        agent_id: str = "default",
        title: str | None = None,
    ) -> dict[str, Any]:
        return await self._client._request(
            "POST",
            "/api/v1/agent/threads",
            scope=scope,
            body={"agentId": agent_id, "title": title},
        )

    async def list(
        self,
        *,
        scope: "PlatosScope | None" = None,
        agent_id: str | None = None,
        limit: int | None = None,
    ) -> list[dict[str, Any]]:
        qs = []
        if agent_id:
            qs.append(f"agentId={agent_id}")
        if limit:
            qs.append(f"limit={limit}")
        suffix = ("?" + "&".join(qs)) if qs else ""
        res = await self._client._request(
            "GET", f"/api/v1/agent/threads{suffix}", scope=scope
        )
        if not res:
            return []
        return list(res.get("threads", []))

    async def get(
        self, thread_id: str, scope: "PlatosScope | None" = None
    ) -> dict[str, Any] | None:
        try:
            return await self._client._request(
                "GET", f"/api/v1/agent/threads/{thread_id}", scope=scope
            )
        except PlatosNotFoundError:
            return None

    async def messages(
        self, thread_id: str, scope: "PlatosScope | None" = None
    ) -> list[dict[str, Any]]:
        res = await self._client._request(
            "GET", f"/api/v1/agent/threads/{thread_id}/messages", scope=scope
        )
        if not res:
            return []
        return list(res.get("messages", []))

    async def artifacts(
        self, thread_id: str, scope: "PlatosScope | None" = None
    ) -> list[dict[str, Any]]:
        res = await self._client._request(
            "GET", f"/api/v1/agent/threads/{thread_id}/artifacts", scope=scope
        )
        if not res:
            return []
        return list(res.get("artifacts", []))

    async def archive(
        self, thread_id: str, scope: "PlatosScope | None" = None
    ) -> None:
        await self._client._request(
            "POST", f"/api/v1/agent/threads/{thread_id}/archive", scope=scope
        )

    async def unarchive(
        self, thread_id: str, scope: "PlatosScope | None" = None
    ) -> None:
        await self._client._request(
            "POST", f"/api/v1/agent/threads/{thread_id}/unarchive", scope=scope
        )

    async def delete(
        self, thread_id: str, scope: "PlatosScope | None" = None
    ) -> None:
        await self._client._request(
            "DELETE", f"/api/v1/agent/threads/{thread_id}", scope=scope
        )

    # ---- realtime ----

    async def send(
        self,
        thread_id: str,
        message: str,
        *,
        scope: "PlatosScope | None" = None,
        agent_id: str | None = None,
        context_type: str | None = None,
        context_id: str | None = None,
        dynamic_blocks: dict[str, str] | None = None,
        attachment_ids: list[str] | None = None,
        max_reconnect_attempts: int = _DEFAULT_MAX_RECONNECT,
    ) -> AsyncGenerator[dict[str, Any], None]:
        """Send a message on a thread and stream events back.

        The agent service speaks Socket.IO on the `/agent` namespace; we
        use the raw WebSocket transport here so we don't have to pull in
        a Socket.IO Python client. Payload framing matches Socket.IO v4
        EIO=4 packets: `40` (open namespace), `42["event", {...}]` (send).
        """
        url = self._ws_url()
        sent_initial = False
        reconnect_attempt = 0

        async def _emit_ready(ws: ClientConnection) -> None:
            # Socket.IO EIO=4 "connect" frame to the /agent namespace.
            await ws.send(f"40{self._client.socket_namespace},")

        async def _emit_message(ws: ClientConnection) -> None:
            payload: dict[str, Any] = {
                "message": message,
                "threadId": thread_id,
            }
            if agent_id:
                payload["agentId"] = agent_id
            if context_type:
                payload["contextType"] = context_type
            if context_id:
                payload["contextId"] = context_id
            if dynamic_blocks:
                payload["dynamicBlocks"] = dynamic_blocks
            if attachment_ids:
                payload["attachmentIds"] = attachment_ids
            frame = f'42{self._client.socket_namespace},["message",{json.dumps(payload)}]'
            await ws.send(frame)

        async def _emit_resume(ws: ClientConnection) -> None:
            frame = (
                f'42{self._client.socket_namespace},'
                f'["resume_stream",{json.dumps({"threadId": thread_id})}]'
            )
            await ws.send(frame)

        while True:
            try:
                async with websockets.connect(
                    url,
                    additional_headers=self._ws_auth_headers(scope),
                ) as ws:
                    yield {"type": "connected"}
                    await _emit_ready(ws)
                    if not sent_initial:
                        await _emit_message(ws)
                        sent_initial = True
                    else:
                        await _emit_resume(ws)
                    reconnect_attempt = 0
                    async for raw in ws:
                        event = _parse_socketio_frame(raw)
                        if event is None:
                            continue
                        yield event
                        if event.get("type") == "done":
                            return
            except asyncio.CancelledError:
                yield {"type": "done", "stopped": True}
                raise
            except (websockets.ConnectionClosed, websockets.InvalidHandshake) as exc:
                yield {"type": "disconnected", "reason": str(exc)}
            except OSError as exc:
                yield {"type": "disconnected", "reason": str(exc)}

            if reconnect_attempt >= max_reconnect_attempts:
                raise PlatosNetworkError(
                    RuntimeError(
                        f"platos-client: exhausted {max_reconnect_attempts} reconnection attempts"
                    )
                )
            reconnect_attempt += 1
            delay = min(_RECONNECT_MAX_S, _RECONNECT_BASE_S * (2 ** (reconnect_attempt - 1)))
            yield {"type": "reconnecting", "attempt": reconnect_attempt}
            await asyncio.sleep(delay)

    def _ws_url(self) -> str:
        base = self._client.base_url
        if base.startswith("https://"):
            return "wss://" + base[len("https://") :] + "/socket.io/?EIO=4&transport=websocket"
        if base.startswith("http://"):
            return "ws://" + base[len("http://") :] + "/socket.io/?EIO=4&transport=websocket"
        return base + "/socket.io/?EIO=4&transport=websocket"

    def _ws_auth_headers(self, scope: "PlatosScope | None") -> dict[str, str]:
        # Socket.IO auth is normally passed in the `auth` payload; the
        # agent gateway also accepts scope + token as WS headers which
        # lets us stay transport-agnostic without parsing handshake packets.
        return self._client._build_headers(scope)


def _parse_socketio_frame(raw: str | bytes) -> dict[str, Any] | None:
    """Parse a single Socket.IO v4 EIO=4 frame into the wrapped event."""
    text = raw.decode("utf-8") if isinstance(raw, (bytes, bytearray)) else raw
    if not text:
        return None
    # Protocol frames: "0" open / "2" ping / "3" pong / "40" connect ack /
    # "42" event. We only care about "42" (event) and "44" (connect-error).
    if text.startswith("42"):
        # strip namespace: "42/agent,[..]" → "[..]"
        comma = text.find(",", 2)
        if comma == -1:
            body = text[2:]
        else:
            body = text[comma + 1 :]
        try:
            parsed = json.loads(body)
        except json.JSONDecodeError:
            return None
        if isinstance(parsed, list) and parsed and parsed[0] == "agent_event":
            event = parsed[1] if len(parsed) > 1 else {}
            if isinstance(event, dict):
                return event
    if text.startswith("44"):
        return {"type": "error", "message": "namespace connect error"}
    return None

File: packages/platos-client/src/apis/threads.ts
/**
 * @platosdev/client — threads API + realtime streaming. Theme I.1 + I.3.
 *
 * `send()` is the important one — returns an async-iterable of
 * `PlatosStreamEvent`. Hardening (I.3):
 *
 *   - Manual reconnection loop with exponential backoff. We set
 *     `reconnection: false` on the underlying socket so we can own the
 *     state transitions precisely.
 *   - Events that arrive while we're reconnecting are buffered — never
 *     dropped. `{ type: "reconnecting", retryCount }` is surfaced to the
 *     caller so UIs can show a banner.
 *   - Abort signal tears down the socket and emits a terminal
 *     `{ type: "done", stopped: true }`.
 */

import type { Socket } from "socket.io-client";
import type { PlatosClient } from "../client.js";
import { PlatosNotFoundError } from "../errors.js";
import type {
  PlatosArtifact,
  PlatosMessage,
  PlatosScope,
  PlatosStreamEvent,
  PlatosThread,
  SendMessageOptions,
} from "../types.js";

const DEFAULT_MAX_RECONNECT_RETRIES = 5;
const RECONNECT_BASE_DELAY_MS = 500;
const RECONNECT_MAX_DELAY_MS = 10_000;

export class ThreadsApi {
  constructor(private readonly client: PlatosClient) {}

  /**
   * Create a new conversation thread. The server mints the id + stores
   * the scope tuple; the returned `thread.id` is what you hand to
   * `send()`.
   */
  async create(
    scope?: PlatosScope,
    options: { agentId?: string; title?: string } = {},
  ): Promise<PlatosThread> {
    return this.client._fetch<PlatosThread>(
      "/api/v1/agent/threads",
      {
        method: "POST",
        body: JSON.stringify({
          agentId: options.agentId ?? "default",
          title: options.title ?? null,
        }),
      },
      scope,
    );
  }

  /** List threads in the current scope. */
  async list(scope?: PlatosScope, query?: { agentId?: string; limit?: number }): Promise<PlatosThread[]> {
    const qs = new URLSearchParams();
    if (query?.agentId) qs.set("agentId", query.agentId);
    if (query?.limit) qs.set("limit", String(query.limit));
    const suffix = qs.toString() ? `?${qs.toString()}` : "";
    const res = await this.client._fetch<{ threads: PlatosThread[] }>(
      `/api/v1/agent/threads${suffix}`,
      { method: "GET" },
      scope,
    );
    return res?.threads ?? [];
  }

  async get(threadId: string, scope?: PlatosScope): Promise<PlatosThread | null> {
    try {
      return await this.client._fetch<PlatosThread>(
        `/api/v1/agent/threads/${encodeURIComponent(threadId)}`,
        { method: "GET" },
        scope,
      );
    } catch (err) {
      if (err instanceof PlatosNotFoundError) return null;
      throw err;
    }
  }

  /** Fetch messages for a thread (F.x paginated). */
  async messages(threadId: string, scope?: PlatosScope): Promise<PlatosMessage[]> {
    const res = await this.client._fetch<{ messages: PlatosMessage[] }>(
      `/api/v1/agent/threads/${encodeURIComponent(threadId)}/messages`,
      { method: "GET" },
      scope,
    );
    return res?.messages ?? [];
  }

  /** Fetch artifacts attached to a thread (F.8). */
  async artifacts(threadId: string, scope?: PlatosScope): Promise<PlatosArtifact[]> {
    const res = await this.client._fetch<{ artifacts: PlatosArtifact[] }>(
      `/api/v1/agent/threads/${encodeURIComponent(threadId)}/artifacts`,
      { method: "GET" },
      scope,
    );
    return res?.artifacts ?? [];
  }

  /** Archive / unarchive / pin toggles. */
  async archive(threadId: string, scope?: PlatosScope): Promise<void> {
    await this.client._fetch<void>(
      `/api/v1/agent/threads/${encodeURIComponent(threadId)}/archive`,
      { method: "POST" },
      scope,
    );
  }

  async unarchive(threadId: string, scope?: PlatosScope): Promise<void> {
    await this.client._fetch<void>(
      `/api/v1/agent/threads/${encodeURIComponent(threadId)}/unarchive`,
      { method: "POST" },
      scope,
    );
  }

  async delete(threadId: string, scope?: PlatosScope): Promise<void> {
    await this.client._fetch<void>(
      `/api/v1/agent/threads/${encodeURIComponent(threadId)}`,
      { method: "DELETE" },
      scope,
    );
  }

  /**
   * Send a message to an agent on a thread and stream the response back
   * as an async iterator of `PlatosStreamEvent`s. The iterator completes
   * when the server emits `{ type: "done" }`, when the caller's
   * `signal` aborts, or when reconnection budget is exhausted.
   */
  async *send(
    threadId: string,
    message: string,
    options: SendMessageOptions = {},
    scope?: PlatosScope,
  ): AsyncGenerator<PlatosStreamEvent, void, void> {
    const maxReconnect = options.maxReconnectRetries ?? DEFAULT_MAX_RECONNECT_RETRIES;
    const queue: PlatosStreamEvent[] = [];
    let waker: (() => void) | null = null;
    let closed = false;
    let terminalError: Error | null = null;

    // Single shared wake helper — resolving the current waker lets the
    // yield loop pop `queue` or observe `closed`.
    const wake = () => {
      if (waker) {
        const w = waker;
        waker = null;
        w();
      }
    };
    const push = (ev: PlatosStreamEvent) => {
      queue.push(ev);
      wake();
    };

    let currentSocket: Socket | null = null;
    let reconnectRetryCount = 0;

    const emitMessage = (sock: Socket) =>
      sock.emit("message", {
        message,
        threadId,
        agentId: options.agentId,
        contextType: options.contextType,
        contextId: options.contextId,
        dynamicBlocks: options.dynamicBlocks,
        attachmentIds: options.attachmentIds,
        sessionContextOverride: options.sessionContextOverride,
        // Per-request model routing: caller selects a named route on the agent.
        ...(options.modelLabel ? { modelLabel: options.modelLabel } : {}),
      });

    // After a reconnect we tell the server which thread we want to
    // resume streaming for. The agent gateway maintains per-thread
    // rooms so it can replay any events we missed during the gap.
    const emitResume = (sock: Socket) =>
      sock.emit("resume_stream", { threadId });

    const attachSocketHandlers = (sock: Socket) => {
      sock.on("agent_event", (ev: PlatosStreamEvent) => {
        push(ev);
        if (ev.type === "done") {
          closed = true;
          sock.disconnect();
          wake();
        }
      });
      sock.on("error", (err: unknown) => {
        // Transport-level error. Treat like a disconnect so we retry
        // reconnection instead of failing hard — many fly/caddy setups
        // surface a transient "error" before "disconnect".
        push({ type: "error", message: err instanceof Error ? err.message : String(err) });
      });
      sock.on("disconnect", (reason: string) => {
        push({ type: "disconnected", reason });
        if (closed) {
          wake();
          return;
        }
        // Kick off reconnect unless the caller aborted or the reason is
        // explicitly client-side (e.g. we called `.disconnect()`).
        if (reason === "io client disconnect" || options.signal?.aborted) {
          wake();
          return;
        }
        void scheduleReconnect();
      });
    };

    const scheduleReconnect = async () => {
      if (closed) return;
      if (reconnectRetryCount >= maxReconnect) {
        terminalError = new Error(`platos-client: exhausted ${maxReconnect} reconnection retries`);
        closed = true;
        wake();
        return;
      }
      reconnectRetryCount += 1;
      const delay = Math.min(RECONNECT_MAX_DELAY_MS, RECONNECT_BASE_DELAY_MS * Math.pow(2, reconnectRetryCount - 1));
      push({ type: "reconnecting", retryCount: reconnectRetryCount });
      await new Promise<void>((resolve) => setTimeout(resolve, delay));
      if (closed || options.signal?.aborted) return;
      try {
        currentSocket = await this.client._openSocketWithRefresh(scope);
        attachSocketHandlers(currentSocket);
        push({ type: "connected" });
        emitResume(currentSocket);
      } catch (err) {
        terminalError = err instanceof Error ? err : new Error(String(err));
        closed = true;
        wake();
      }
    };

    // Open the first socket through the same refresh-capable auth path as REST.
    currentSocket = await this.client._openSocketWithRefresh(scope);
    attachSocketHandlers(currentSocket);
    push({ type: "connected" });
    emitMessage(currentSocket);

    options.signal?.addEventListener("abort", () => {
      push({ type: "done", stopped: true });
      closed = true;
      currentSocket?.disconnect();
      wake();
    });

    try {
      while (!closed || queue.length > 0) {
        if (queue.length > 0) {
          yield queue.shift()!;
          continue;
        }
        if (terminalError) throw terminalError;
        if (closed) break;
        await new Promise<void>((resolve) => {
          waker = resolve;
        });
      }
    } finally {
      currentSocket?.disconnect();
    }
  }
}
